File: twisted/server-04-translate.py
#!/usr/bin/env python
from twisted.web import server, resource
from twisted.internet import reactor
from twisted.web.server import NOT_DONE_YET

from utils import validate_params

SVC_HOST = 'localhost'
SVC_PORT = 8010

PORT = 8080

class TranslatorResource(resource.Resource):
    isLeaf = True

    def render_GET(self, request):
        # Ensure URL params were passed
        error_str = validate_params(request, ('data',))
        if error_str: return error_str

        from txtranslator import TranslatorClient
        client = TranslatorClient(SVC_HOST, SVC_PORT)
        # Pass callback function only (Not doing any error handling)
        client.translate1(request.args['data'][0], translator_callback, request)

        return NOT_DONE_YET

def translator_callback(result, request):
    request.write(result)
    request.finish()

root = resource.Resource()
root.putChild("translate", TranslatorResource())
site = server.Site(root)
reactor.listenTCP(PORT, site)
reactor.run()

print "Listening on http://localhost:%s/translate/ params:data" % PORT

File: nodejs/server-04-translate.js
var http = require("http");
var url = require('url');
var net = require('net');
var translator = require('./ndtranslator');

var SVC_HOST = 'localhost';
var SVC_PORT = 8010;

var PORT = 8080

http.createServer(function(request, response) {
    if (request.url == '/favicon.ico'){
        response.statusCode = 404;
        response.end();
        return;
    }
    var parts = url.parse(request.url, true);

    translator.translate(SVC_HOST, SVC_PORT, parts.query.data, function(err, output_str){
        if (err == null){
            response.write(output_str);
        } else {
            response.statusCode = 500;
            response.write("Internal Server Error: " + err);
        }
        response.end();
    });

}).listen(PORT);

console.log("Listening on http://localhost:" + PORT + " params:data");
